### PR TITLE
Facilitate addition of new pod/*.pod file

### PR DESCRIPTION
**Pull request by jkeenan**
For: https://github.com/Perl/perl5/issues/18413

### CODE REVIEW

File: Porting/add-pod-file [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+        push @after_section, $l;
+        $this++;
+    }
+    elsif (! $this) {
+        push @target_section, $l;
+    }
+    else {
+        push @after_section, $l;
+    }
+}
```
Comment (rjbs) [inline Porting/add-pod-file:217]: I found this section confusing, although after a lot of slow re-reading, I think it is correct.  Part of the problem is that I would have tackled this whole problem in a different, but not necessarily better, way.  But in this one section (splitting @balance into "target" and "after") I think I would suggest at minimum two small changes.

1. eliminate `$this`, which is a fairly meaningless name for something with an important meaning
2. replace the if/elsif/else with a (two-clause) if/else

I believe that would replace the code from L201-213 with:

```perl
my $target = \@target_section;
for my $l (@balance) {
    $target = \@after_section if $l =~ m/^=(head2|for)/;
    push @$target, $l;
}
```

The reassignment of $target within the loop will happen many times, but not so many that the performance cost will be noticed, especially given that this program will be run once or twice each year.

File: Porting/add-pod-file [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+    $man_sections_str .= "\t$man_sections[$i] => $man_sections[$j]\n";
+}
+my %man_sections_seen = map { $_ => 1 } @man_section_abbrevs;
+my $man_sections = { @man_sections };
+
+my ($stub, $abstract, $section, $verbose) = ('') x 4;
+GetOptions(
+    "stub=s"        => \$stub,
+    "abstract=s"    => \$abstract,
+    "section=s"     => \$section,
```
Comment (rjbs) [inline Porting/add-pod-file:120]: Above, under "well-formatted" you suggest that the pod file being added must have a `NAME` section with a stub and abstract, but here you require them in the program.  This seems like more work for the human running this program, as well as an opportunity to create disagreement.  I think you should extract the stub and abstract from the source file, and you should eliminate these options.

**Review by rjbs — CHANGES_REQUESTED** [review]
I think this looks good!  I have made two changes, one of which addresses ease of use and error prevention, and one of which is about (my take on) code clarity.

File: Porting/add-pod-file (hunk 2 above)
Comment (jkeenan) [reply Porting/add-pod-file:120]: > Above, under "well-formatted" you suggest that the pod file being added must have a `NAME` section with a stub and abstract, but here you require them in the program. This seems like more work for the human running this program, as well as an opportunity to create disagreement. I think you should extract the stub and abstract from the source file, and you should eliminate these options.

Yes, that was on mental TODO list and is the main reason I characterized this as a *first* pass.  Do you know if we have any prior art for parsing of a F<.pod> file to abstract what I call the 'stub' and the abstract?

I'll look at your other suggestions over the next several days.

Thank you very much.
Jim Keenan

Comment (Grinnz) [reply Porting/add-pod-file:120]: I don't know of prior art here but it should be parsed the same way as manpage indexers do.

one, or, more, comma-separated, names - rest of paragraph is the abstract

Comment (Leont) [reply Porting/add-pod-file:120]: > Yes, that was on mental TODO list and is the main reason I characterized this as a first pass. Do you know if we have any prior art for parsing of a F<.pod> file to abstract what I call the 'stub' and the abstract?

Module::Metadata can be helpful here:
```
my $data = Module::Metadata->new_from_file($filename, collect_pod => 1, decode_pod => 1);
		
my ($abstract) = ($data->pod('NAME') // '')  =~ / \A \s+ \S+ \s? - \s? (.+?) \s* \z /x or warn "Could not parse abstract from `=head1 NAME` in $filename";
```
(Example taken from [App::ModuleBuildTiny::Dist](https://github.com/Leont/app-modulebuildtiny/blob/master/lib/App/ModuleBuildTiny/Dist.pm#L147))

Comment (Grinnz) [reply Porting/add-pod-file:120]: Comments below
```
/ \A 
  \s+ # should be \s*
  \S+ # should allow for multiple names separated by comma and space
  \s? # should be \s+ - space before and after hyphen is required to disambiguate from names with hyphens in them
  - 
  \s? # ditto above
  (.+?) \s* \z /x
```

Comment (Grinnz) [reply Porting/add-pod-file:120]: Here is the regex dzil uses: https://metacpan.org/source/RJBS/Dist-Zilla-6.017/lib/Dist/Zilla/Util.pm#L54

**Comment by jkeenan** [comment]
Have pushed additional commits.  I tested the program with all possible values for `--section`.  They all worked except for `P` (platform-specific), in which case for some reason `win32/pod.mak` was not being modified.  I noted that bug in the documentation, but I don't think it's serious enough to prevent this from going into blead now.

Thank you very much.
Jim Keenan

File: Porting/add-pod-file [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+croak "Cannot locate 'podchecker' within this checkout; have you called 'make'?"
+    unless -f $thispodchecker;
+
+say "Step 1: Basic test of validity of POD in $newpodpath" if $verbose;
+
+system(qq|$^X $thispodchecker $newpodpath|)
+    and croak "$newpodpath has POD errors; correct before proceeding further";
+my $data = Module::Metadata->new_from_file($newpodpath, collect_pod => 1, decode_pod => 1);
+
+my ($abstract) = ($data->pod('NAME') // '')  =~ / \A \s+ \S+ \s? - \s? (.+?) \s* \z /x
```
Comment (Grinnz) [inline Porting/add-pod-file]: This regex is too restrictive as I pointed out in my further comments

Comment (jkeenan) [reply Porting/add-pod-file]: I believe that for the purpose of this program, Leon's regex is sufficient.  It handles F<.pod> files with C<NAME> sections composed in the manner in which the overwhelming majority of existing F<pod/\*.pod> files are composed.  If you can show me a potential new F<pod/*.pod> which needs your regex pattern, I will accept it.  Otherwise, I will not.

Thank you very much.
Jim Keenan

Comment (Grinnz) [reply Porting/add-pod-file]: Three examples:

Some modules like https://perldoc.perl.org/Encode::Supported have a double-dash before the abstract - this is arguably wrong but the dzil regex handles this.

Any documentation for a script containing a hyphen may be misparsed by that, since it doesn't require spaces before and after the hyphen.

Any documentation such as https://perldoc.perl.org/diagnostics with multiple program names will be misparsed.

Comment (Grinnz) [reply Porting/add-pod-file]: Additionally I will add - I would prefer the abstract parsing code to be more correct than less for the sake of future cargo culting :)

**Comment by rjbs** [comment]
Hey, circling back after a wonky week of work.

What @jkeenan has written looks pretty good to me, but I agree that the regex could be tweaked slightly to make it work.  The perfect shouldn't be the enemy of the good enough, but is this just "replace one regex with another" and we're good?  

@Grinnz If so, could you provide a "suggestion"-style comment that we could just apply?  Failing that, let's let Jim apply this, then open a PR for the improvement.  I can help get that in place while Jim can rest easy knowing that even if you and I fall off the planet, he has gotten something useful enough into core.

File: Porting/add-pod-file (hunk 3 above)
Comment (Grinnz) [inline Porting/add-pod-file]: Lightly edited from https://metacpan.org/source/RJBS/Dist-Zilla-6.017/lib/Dist/Zilla/Util.pm#L54 as previously suggested.
```suggestion
my ($abstract) = ($data->pod('NAME') // '')  =~ /\A(?:\S+\s+)+?-+\s+(.+?)\s*\z/s
```

Comment (rjbs) [reply Porting/add-pod-file]: Thanks, Dan.  @jkeenan I think this change looks very slightly better, and I think you should be able to just click "commit suggestion" and everything will be cool.  Let me know if I am mistaken. Today has had a fair bit of that already… ;)

Comment (Leont) [reply Porting/add-pod-file]: Has that been tested? AFAIK that that whitespace before the name is actually essential. Module::Metadata->pod will include the empty line between the `=head` and the actual content (I'm guessing that's stripped off before it hits the regex in dzil's case).

Comment (Leont) [reply Porting/add-pod-file]: I guess the better regex for the first part would just be `[^-]+`. Maybe something like `/ \A [^-]+? \s+ -+ \s+ (.+?) \s* \z /x`

Comment (jkeenan) [reply Porting/add-pod-file]: Here is the sample file I used in many hours of testing:
```
$ cat perlphonypod.pod 
=encoding utf8

=head1 NAME

perlphonypod - Phony POD

=head1 DESCRIPTION

This document is nonsense.

=cut

```
Here is a program I wrote to test @LeonT's and @Grinnz's suggested regexes against this sample file:
```
$ cat test-regex.pl 
use 5.14.0;
use warnings;
use Module::Metadata;

my ($file, $data, $abstract);
$file = 'perlphonypod.pod';
$data = Module::Metadata->new_from_file($file, collect_pod => 1, decode_pod => 1);

say "Leon's original suggestion:";
($abstract) = ($data->pod('NAME') // '')  =~ / \A \s+ \S+ \s? - \s? (.+?) \s* \z /x
    or warn "Could not parse abstract from `=head1 NAME` in $file";
say "<$abstract>";
say "";

say "Dan's suggestion based on Dist-Zilla:";
($abstract) = ($data->pod('NAME') // '')  =~ /\A(?:\S+\s+)+?-+\s+(.+?)\s*\z/s
    or warn "Could not parse abstract from `=head1 NAME` in $file";
```
Let's run it.
```
$ perl test-regex.pl 
Leon's original suggestion:
<Phony POD>

Dan's suggestion based on Dist-Zilla:
Could not parse abstract from `=head1 NAME` in perlphonypod.pod at test-regex.pl line 16.
```
@LeonT's suggestion works; @Grinnz's does not.

I next adapted the program above into a new program which I ran against the existing `pod/*.pod` files, using @LeonT's suggested regex:
```
$ cat xtest-regex.pl 
use 5.14.0;
use warnings;
use Module::Metadata;
#use Data::Dump qw(dd pp);
use Carp;
use File::Spec;

chdir $ENV{PERL_WORKDIR} or die;

my @podfiles = ();
opendir my $DIRH, 'pod' or croak "Unable to opendir";
@podfiles = sort map { File::Spec->catfile('pod', $_) } grep { ! /^\.{1,2}$/ } grep { /\.pod$/ } readdir($DIRH);
closedir $DIRH or croak "Unable to closedir";
#dd(\@podfiles);

my $regex = qr/ \A \s+ \S+ \s? - \s? (.+?) \s* \z /x;

my $goodcount = 0;
my @nonmatching = ();

for my $p (@podfiles) {
    next if $p =~ /perl5336delta/ or $p =~ m/perldelta/; # This is 'in progress'
    next if $p =~ /perltoc/; # non-standard
    #say "Working on $p";
    my ($data, $abstract);
    eval {
        local $@;
        $data = Module::Metadata->new_from_file($p, collect_pod => 1, decode_pod => 1);
        if ($@) {
            say "XXX: new_from_file() failed on $p: $@";
        }
    };

    if (($abstract) = ($data->pod('NAME') // '')  =~ $regex) {
        say "$p - $abstract";
        $goodcount++;
    }
    else {
        push @nonmatching, $p;
    }
}
say "Count matching regex:     $goodcount";
say "Count NOT matching regex: ", scalar @nonmatching;
say "  $_" for @nonmatching;
for my $p (@nonmatching) {
    system(qq|ack -A4 '^=head1 NAME' $p|)
       and carp "Unable to grep $p";
}
```
Let's run it.
```
$ perl xtest-regex.pl
pod/perl.pod - The Perl 5 language interpreter
pod/perl5004delta.pod - what's new for perl5.004
pod/perl5005delta.pod - what's new for perl5.005
pod/perl5100delta.pod - what is new for perl 5.10.0
pod/perl5101delta.pod - what is new for perl v5.10.1
pod/perl5120delta.pod - what is new for perl v5.12.0
pod/perl5121delta.pod - what is new for perl v5.12.1
pod/perl5122delta.pod - what is new for perl v5.12.2
pod/perl5123delta.pod - what is new for perl v5.12.3
pod/perl5124delta.pod - what is new for perl v5.12.4
pod/perl5125delta.pod - what is new for perl v5.12.5
pod/perl5140delta.pod - what is new for perl v5.14.0
pod/perl5141delta.pod - what is new for perl v5.14.1
pod/perl5142delta.pod - what is new for perl v5.14.2
pod/perl5143delta.pod - what is new for perl v5.14.3
pod/perl5144delta.pod - what is new for perl v5.14.4
pod/perl5160delta.pod - what is new for perl v5.16.0
pod/perl5161delta.pod - what is new for perl v5.16.1
pod/perl5162delta.pod - what is new for perl v5.16.2
pod/perl5163delta.pod - what is new for perl v5.16.3
pod/perl5180delta.pod - what is new for perl v5.18.0
pod/perl5181delta.pod - what is new for perl v5.18.1
pod/perl5182delta.pod - what is new for perl v5.18.2
pod/perl5184delta.pod - what is new for perl v5.18.4
pod/perl5200delta.pod - what is new for perl v5.20.0
pod/perl5201delta.pod - what is new for perl v5.20.1
pod/perl5202delta.pod - what is new for perl v5.20.2
pod/perl5203delta.pod - what is new for perl v5.20.3
pod/perl5220delta.pod - what is new for perl v5.22.0
pod/perl5221delta.pod - what is new for perl v5.22.1
pod/perl5222delta.pod - what is new for perl v5.22.2
pod/perl5223delta.pod - what is new for perl v5.22.3
pod/perl5224delta.pod - what is new for perl v5.22.4
pod/perl5240delta.pod - what is new for perl v5.24.0
pod/perl5241delta.pod - what is new for perl v5.24.1
pod/perl5242delta.pod - what is new for perl v5.24.2
pod/perl5243delta.pod - what is new for perl v5.24.3
pod/perl5244delta.pod - what is new for perl v5.24.4
pod/perl5260delta.pod - what is new for perl v5.26.0
pod/perl5261delta.pod - what is new for perl v5.26.1
pod/perl5262delta.pod - what is new for perl v5.26.2
pod/perl5263delta.pod - what is new for perl v5.26.3
pod/perl5280delta.pod - what is new for perl v5.28.0
pod/perl5281delta.pod - what is new for perl v5.28.1
pod/perl5282delta.pod - what is new for perl v5.28.2
pod/perl5283delta.pod - what is new for perl v5.28.3
pod/perl5300delta.pod - what is new for perl v5.30.0
pod/perl5301delta.pod - what is new for perl v5.30.1
pod/perl5302delta.pod - what is new for perl v5.30.2
pod/perl5303delta.pod - what is new for perl v5.30.3
pod/perl5320delta.pod - what is new for perl v5.32.0
pod/perl5330delta.pod - what is new for perl v5.33.0
pod/perl5331delta.pod - what is new for perl v5.33.1
pod/perl5332delta.pod - what is new for perl v5.33.2
pod/perl5333delta.pod - what is new for perl v5.33.3
pod/perl5334delta.pod - what is new for perl v5.33.4
pod/perl5335delta.pod - what is new for perl v5.33.5
pod/perl561delta.pod - what's new for perl v5.6.1
pod/perl56delta.pod - what's new for perl v5.6.0
pod/perl581delta.pod - what is new for perl v5.8.1
pod/perl582delta.pod - what is new for perl v5.8.2
pod/perl583delta.pod - what is new for perl v5.8.3
pod/perl584delta.pod - what is new for perl v5.8.4
pod/perl585delta.pod - what is new for perl v5.8.5
pod/perl586delta.pod - what is new for perl v5.8.6
pod/perl587delta.pod - what is new for perl v5.8.7
pod/perl588delta.pod - what is new for perl v5.8.8
pod/perl589delta.pod - what is new for perl v5.8.9
pod/perl58delta.pod - what is new for perl v5.8.0
pod/perlaix.pod - Perl version 5 on IBM AIX (UNIX) systems
pod/perlamiga.pod - Perl under AmigaOS 4.1
pod/perlandroid.pod - Perl under Android
pod/perlapi.pod - autogenerated documentation for the perl public API
pod/perlapio.pod - perl's IO abstraction interface.
pod/perlartistic.pod - the Perl Artistic License
pod/perlbook.pod - Books about and related to Perl
pod/perlboot.pod - Links to information on object-oriented programming in Perl
pod/perlbot.pod - Links to information on object-oriented programming in Perl
pod/perlcall.pod - Perl calling conventions from C
pod/perlcheat.pod - Perl 5 Cheat Sheet
pod/perlclib.pod - Internal replacements for standard C library functions
Wide character in say at xtest-regex.pl line 35.
pod/perlcn.pod - 简体中文 Perl 指南
pod/perlcommunity.pod - a brief overview of the Perl community
pod/perlcygwin.pod - Perl for Cygwin
pod/perldata.pod - Perl data types
pod/perldbmfilter.pod - Perl DBM Filters
pod/perldebguts.pod - Guts of Perl debugging
pod/perldebtut.pod - Perl debugging tutorial
pod/perldeprecation.pod - list Perl deprecations
pod/perldiag.pod - various Perl diagnostics
pod/perldocstyle.pod - A style guide for writing Perl's documentation
pod/perldos.pod - Perl under DOS, W31, W95.
pod/perldtrace.pod - Perl's support for DTrace
pod/perlebcdic.pod - Considerations for running Perl on EBCDIC platforms
pod/perlembed.pod - how to embed perl in your C program
pod/perlexperiment.pod - A listing of experimental features in Perl
pod/perlfilter.pod - Source Filters
pod/perlfork.pod - Perl's fork() emulation
pod/perlfreebsd.pod - Perl version 5 on FreeBSD systems
pod/perlgit.pod - Detailed information about git and the Perl repository
pod/perlgov.pod - Perl Rules of Governance
pod/perlgpl.pod - the GNU General Public License, version 1
pod/perlguts.pod - Introduction to the Perl API
pod/perlhack.pod - How to hack on Perl
pod/perlhacktips.pod - Tips for Perl core C code hacking
pod/perlhacktut.pod - Walk through the creation of a simple C code patch
pod/perlhaiku.pod - Perl version 5.10+ on Haiku
pod/perlhist.pod - the Perl history records
pod/perlhpux.pod - Perl version 5 on Hewlett-Packard Unix (HP-UX) systems
pod/perlhurd.pod - Perl version 5 on Hurd
pod/perlinterp.pod - An overview of the Perl interpreter
pod/perlintro.pod - a brief introduction and overview of Perl
pod/perliol.pod - C API for Perl's implementation of IO in Layers.
pod/perlipc.pod - Perl interprocess communication (signals, fifos, pipes, safe subprocesses, sockets, and semaphores)
pod/perlirix.pod - Perl version 5 on Irix systems
Wide character in say at xtest-regex.pl line 35.
pod/perljp.pod - 日本語 Perl ガイド
Wide character in say at xtest-regex.pl line 35.
pod/perlko.pod - 한국어 Perl 안내서
pod/perllexwarn.pod - Perl Lexical Warnings
pod/perllinux.pod - Perl version 5 on Linux systems
pod/perllocale.pod - Perl locale handling (internationalization and localization)
pod/perllol.pod - Manipulating Arrays of Arrays in Perl
pod/perlmacos.pod - Perl under Mac OS (Classic)
pod/perlmacosx.pod - Perl under Mac OS X
pod/perlmod.pod - Perl modules (packages and symbol tables)
pod/perlmodinstall.pod - Installing CPAN Modules
pod/perlmodlib.pod - constructing new Perl modules and finding existing ones
pod/perlmodstyle.pod - Perl module style guide
pod/perlmroapi.pod - Perl method resolution plugin interface
pod/perlnetware.pod - Perl for NetWare
pod/perlnewmod.pod - preparing a new module for distribution
pod/perlnumber.pod - semantics of numbers and numeric operations in Perl
pod/perlootut.pod - Object-Oriented Programming in Perl Tutorial
pod/perlopenbsd.pod - Perl version 5 on OpenBSD systems
pod/perlopentut.pod - simple recipes for opening files and pipes in Perl
pod/perlos2.pod - Perl under OS/2, DOS, Win0.3*, Win0.95 and WinNT.
pod/perlos390.pod - building and installing Perl for OS/390 and z/OS
pod/perlpacktut.pod - tutorial on C<pack> and C<unpack>
pod/perlperf.pod - Perl Performance and Optimization Techniques
pod/perlplan9.pod - Plan 9-specific documentation for Perl
pod/perlpodspec.pod - Plain Old Documentation: format specification and notes
pod/perlpodstyle.pod - Perl POD style guide
pod/perlpolicy.pod - Various and sundry policies and commitments related to the Perl core
pod/perlport.pod - Writing portable Perl
pod/perlpragma.pod - how to write a user pragma
pod/perlqnx.pod - Perl version 5 on QNX
pod/perlreapi.pod - Perl regular expression plugin interface
pod/perlrebackslash.pod - Perl Regular Expression Backslash Sequences and Escapes
pod/perlreftut.pod - Mark's very short tutorial about references
pod/perlreguts.pod - Description of the Perl regular expression engine.
pod/perlrepository.pod - Links to current information on the Perl source repository
pod/perlrequick.pod - Perl regular expressions quick start
pod/perlreref.pod - Perl Regular Expressions Reference
pod/perlretut.pod - Perl regular expressions tutorial
pod/perlriscos.pod - Perl version 5 for RISC OS
pod/perlrun.pod - how to execute the Perl interpreter
pod/perlsec.pod - Perl security
pod/perlsecpolicy.pod - Perl security report handling policy
pod/perlsolaris.pod - Perl version 5 on Solaris systems
pod/perlsource.pod - A guide to the Perl source tree
pod/perlstyle.pod - Perl style guide
pod/perlsynology.pod - Perl 5 on Synology DSM systems
pod/perlthrtut.pod - Tutorial on threads in Perl
pod/perltodo.pod - Link to the Perl to-do list
pod/perltooc.pod - Links to information on object-oriented programming in Perl
pod/perltoot.pod - Links to information on object-oriented programming in Perl
pod/perltrap.pod - Perl traps for the unwary
pod/perltru64.pod - Perl version 5 on Tru64 (formerly known as Digital UNIX formerly known as DEC OSF/1) systems
Wide character in say at xtest-regex.pl line 35.
pod/perltw.pod - 正體中文 Perl 指南
pod/perlunicode.pod - Unicode support in Perl
pod/perlunicook.pod - cookbookish examples of handling Unicode in Perl
pod/perlunifaq.pod - Perl Unicode FAQ
pod/perluniintro.pod - Perl Unicode introduction
pod/perluniprops.pod - Index of Unicode Version 13.0.0 character properties in Perl
pod/perlunitut.pod - Perl Unicode Tutorial
pod/perlutil.pod - utilities packaged with the Perl distribution
pod/perlvar.pod - Perl predefined variables
pod/perlvms.pod - VMS-specific documentation for Perl
pod/perlvos.pod - Perl for Stratus OpenVOS
pod/perlwin32.pod - Perl under Windows
Count matching regex:     179
Count NOT matching regex: 16
  pod/perlbs2000.pod
  pod/perldebug.pod
  pod/perldsc.pod
  pod/perlform.pod
  pod/perlfunc.pod
  pod/perlintern.pod
  pod/perlobj.pod
  pod/perlop.pod
  pod/perlos400.pod
  pod/perlpod.pod
  pod/perlre.pod
  pod/perlrecharclass.pod
  pod/perlref.pod
  pod/perlsub.pod
  pod/perlsyn.pod
  pod/perltie.pod
=head1 NAME

perlbs2000 - building and installing Perl for BS2000.

B<This document needs to be updated, but we don't know what it should say.
=head1 NAME
X<debug> X<debugger>

perldebug - Perl debugging

=head1 NAME
X<data structure> X<complex data structure> X<struct>

perldsc - Perl Data Structures Cookbook

=head1 NAME
X<format> X<report> X<chart>

perlform - Perl formats

=head1 NAME
X<function>

perlfunc - Perl builtin functions

=head1 NAME

perlintern - autogenerated documentation of purely B<internal>
Perl functions

=head1 NAME
X<object> X<OOP>

perlobj - Perl object reference

=head1 NAME
X<operator>

perlop - Perl operators and precedence

=head1 NAME

perlos400 - Perl version 5 on OS/400

B<This document needs to be updated, but we don't know what it should say.
=head1 NAME
X<POD> X<plain old documentation>

perlpod - the Plain Old Documentation format

=head1 NAME
X<regular expression> X<regex> X<regexp>

perlre - Perl regular expressions

=head1 NAME
X<character class>

perlrecharclass - Perl Regular Expression Character Classes

=head1 NAME
X<reference> X<pointer> X<data structure> X<structure> X<struct>

perlref - Perl references and nested data structures

=head1 NAME
X<subroutine> X<function>

perlsub - Perl subroutines

=head1 NAME
X<syntax>

perlsyn - Perl syntax

=head1 NAME
X<tie>

perltie - how to hide an object class in a simple variable

```
So, in the overwhelming majority of cases of `pod/*.pod` files which *don't* match @LeonT's regex, the reason is that we have modified these files to include the special `X<*>` syntax.  That's syntax that the author of a proposed new `pod/*.pod` file is likely to add inside the `NAME` section.

Let's keep in mind that our objective here is **not** to find the one perfect regex that matches the `NAME` section of all existing `pod/*.pod` files.  Our objective here is to reduce the amount of work a committer needs to do to get a new such file into blead in the first place.  This pull request will accomplish that as is.

Thank you very much.
Jim Keenan

Comment (Leont) [reply Porting/add-pod-file]: Based on that, I think the winning regex is: `/ ^ [^-]+? \s+ -+ \s+ (.+?) \s* $ /xms`, that seems to catch all of them for me.

Comment (jkeenan) [reply Porting/add-pod-file]: > Based on that, I think the winning regex is: `/ ^ [^-]+? \s+ -+ \s+ (.+?) \s* $ /xms`, that seems to catch all of them for me.

Wow, yes!  Plugging it into my second test program, I get at the end of output:
```
Count matching regex:     195
Count NOT matching regex: 0
```
Thanks!

Comment (jkeenan) [reply Porting/add-pod-file]: I have included Leon's last suggestion in the p.r.

Comment (Grinnz) [reply Porting/add-pod-file]: No, `[^-]+` will not work on programs with hyphens in the name as I mentioned. Simply adding `\s*` at the beginning of the regex I suggested would solve the problem you mentioned.

File: Porting/add-pod-file [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+croak "Cannot locate 'podchecker' within this checkout; have you called 'make'?"
+    unless -f $thispodchecker;
+
+say "Step 1: Basic test of validity of POD in $newpodpath" if $verbose;
+
+system(qq|$^X $thispodchecker $newpodpath|)
+    and croak "$newpodpath has POD errors; correct before proceeding further";
+my $data = Module::Metadata->new_from_file($newpodpath, collect_pod => 1, decode_pod => 1);
+
+my $regex = qr/ ^ [^-]+? \s+ -+ \s+ (.+?) \s* $ /xms;
```
Comment (Grinnz) [inline Porting/add-pod-file]:
```suggestion
my $regex = qr/\A\s*(?:\S+\s+)+?-+\s+(.+?)\s*\z/s;
```

File: Porting/add-pod-file (hunk 3 above)
Comment (Grinnz) [reply Porting/add-pod-file]: And this is also wrong because it will fail to capture the end of an abstract if there is a line break in the paragraph.
